pynhm/base/budget.py

```diff
@@ -0,0 +1,172 @@
+        self.inputs = inputs
+        self.outputs = outputs
+        self.storage_changes = storage_changes
+        self.meta = meta
+        self.verbosity = verbosity
+
+        self._inputs_sum = None
+        self._outputs_sum = None
+        self._storage_changes_sum = None
+        self._balance = None
+        self._accumulations = None
+
+        self.set_initial_accumulations(init_accumulations)
+        return
+
+    @staticmethod
+    def from_storage_unit(storage_unit, **kwargs):
+        # assemble the meta data, which will determine the budget component variables
+        kwargs["meta"] = {}
+        kwargs["meta"]["inputs"] = storage_unit.meta._get_attr_key_val(
+            "input", "var_category", "mass flux"
+        )
+        kwargs["meta"]["outputs"] = storage_unit.meta._get_attr_key_val(
+            "var", "var_category", "mass flux"
+        )
+        kwargs["meta"][
+            "storage_changes"
+        ] = storage_unit.meta._get_attr_key_val(
+            "var", "var_category", "mass storage change"
+        )
+
+        # keep only the desried metadata for budget
+        meta_keys = Budget.get_meta_keys()
+        for component in self.components:
+            for key, val in kwargs["meta"][component].items():
+                kwargs["meta"][component][key] = {
+                    kk: vv for kk, vv in val.items() if kk in meta_keys
+                }
+
+        for component in self.components:
+            kwargs[component] = {}
+            for var in kwargs["meta"][component].keys():
+                kwargs[component][var] = storage_unit[var][:]
+
+        budget = Budget(**kwargs)
+        budget.storage_unit_name = storage_unit.name
+        return budget
+
+    @staticmethod
+    def get_meta_keys():
+        """Return a tuple of the metadata keys used by Budget."""
+        return ("desc", "modules", "var_category", "units")
+
+    @staticmethod
+    def get_components():
+        return ("inputs", "outputs", "storage_changes")
+
+    @property
+    def components(self):
+        return self.get_components()
+
+    def set_initial_accumulations(self, init_accumulations):
+        # init to zero
+        self._accumulations = {}
+        for component in self.components:
+            self._accumulations[component] = {}
+            for var in self[component].keys():
+                self._accumulations[component][var] = (
+                    self[component][var] * zero
+                )
+
+        if init_accumulations is None:
+            return None
+
+        for component in self.components:
+            if component not in init_accumulations.keys():
+                continue
+            for var in self[component].keys():
+                if var in init_accumulations[component].keys():
+                    self._accumulations[component][var] = init_accumulations[
+                        "component"
+                    ][var]
+
+        return None
+
+    def accumulate(self):
```

Comment:
Is this what we do when we "advance" in time?

Reply:
I was trying to keep advance/time out of the budget and just calculate a budget for the timestep it is called... but it's obvious that will not give desired accumulations unless its called with every timestep. So and actual advance should be defined on it.

Reply:
Don't we need the timestep length in order to convert rates to volumes?  I suspect it's in here somewhere but I don't see it.  Or maybe all inflows/outflows/ds are just rates in here and there is no separate tracking of cumulative?  I guess that's the case and accumulate is accumulating inflows and outflows for a time step and not accumulating with time.

Reply:
I would suggest that we keep a running total of cumulative volumes and also store instantaneous rates for the current time step.  We track both in modflow and have found it to be quite useful.

Reply:
my main concern was the init arguments. you think that's reasonable?

Reply:
implemented timestep multiplication for accumulated volumes (or energy, i suppose). this could be cleaned up slightly (getting the time_step from control at some point)

__repr__ shows current instantaneous and accumulated. both reports show time of budget

I implemented a bit of stuff so a budget cant be called more than once per time step (we dont implement this for other physics, ... )

there is a reset method which is mostly untested at the moment

Reply:
I did not implement percent discrepancy as in mf6 table.
I also have not plumbed units or metadata through the budget.